src/sentry/spans/buffer_v2.py

```diff
@@ -0,0 +1,290 @@
+
+                p.execute()
+
+        metrics.timing("sentry.spans.buffer.process_spans.num_spans", len(spans))
+        metrics.timing("sentry.spans.buffer.process_spans.num_is_root_spans", is_root_span_count)
+        metrics.timing("sentry.spans.buffer.process_spans.num_has_root_spans", has_root_span_count)
+
+    def flush_segments(
+        self, now: int, max_segments: int = 0
+    ) -> tuple[int, dict[SegmentId, set[bytes]]]:
+        cutoff = now
+
+        with metrics.timer("sentry.spans.buffer.flush_segments.step1"):
+            with self.client.pipeline(transaction=False) as p:
+                for shard in self.assigned_shards:
+                    key = f"span-buf:q:{shard}"
+                    p.zrangebyscore(
+                        key, 0, cutoff, start=0 if max_segments else None, num=max_segments or None
+                    )
+                    p.zcard(key)
+
+                result = iter(p.execute())
+
+        segment_ids = []
+        queue_sizes = []
+
+        with metrics.timer("sentry.spans.buffer.flush_segments.step2"):
+            with self.client.pipeline(transaction=False) as p:
+                # ZRANGEBYSCORE output
+                for segment_span_ids in result:
+                    # process return value of zrevrangebyscore
+                    for segment_id in segment_span_ids:
+                        # TODO: SSCAN
+                        segment_ids.append(segment_id)
+                        p.smembers(segment_id)
+
+                    # ZCARD output
+                    queue_sizes.append(next(result))
+
+                segments = p.execute()
+
+        for shard_i, queue_size in zip(self.assigned_shards, queue_sizes):
+            metrics.timing(
+                "sentry.spans.buffer.flush_segments.queue_size",
+                queue_size,
+                tags={"shard_i": shard_i},
+            )
+
+        return_segments = {}
+
+        for segment_id, segment in zip(segment_ids, segments):
+            return_segment = set()
+            metrics.timing("sentry.spans.buffer.flush_segments.num_spans_per_segment", len(segment))
+            for payload in segment:
+                return_segment.add(payload)
+
+            return_segments[segment_id] = return_segment
+        metrics.timing("sentry.spans.buffer.flush_segments.num_segments", len(return_segments))
+
+        return sum(queue_sizes), return_segments
+
+    def done_flush_segments(self, segment_ids: dict[SegmentId, set[bytes]]):
+        payload_nums = []
+        metrics.timing("sentry.spans.buffer.done_flush_segments.num_segments", len(segment_ids))
+        with metrics.timer("sentry.spans.buffer.done_flush_segments"):
+            with self.client.pipeline(transaction=False) as p:
+                for segment_id, payloads in segment_ids.items():
+                    hrs_key = b"span-buf:hrs:" + segment_id
+                    p.get(hrs_key)
+                    p.delete(hrs_key)
```

Comment:
`hrs_key` is not described anywhere in this file, as it is created by the redis script. It would be great to describe this somewhere.